<a href="https://colab.research.google.com/github/JairEsc/Mat_Apl_2/blob/main/Biomate_NN_class.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import matplotlib.pyplot as plt
import tqdm
from time import time

# Estructura de los datos.

Consideramos un caso general, una base con datos de frecuencias de especies, cada observación de dimensión $k$, y supongamos que tenemos $n$ observaciones.

El output busca ser la clasificación en una de $d$ categorías, definidas ya sea por ciudad o localización-clima (pero de un sólo tipo).


# Datos de ejemplo

In [2]:
import pandas as pd
import matplotlib.pyplot as plt

df = pd.read_csv("Dataset_w_tags.csv")
df.head()
from sklearn.utils import shuffle
df = shuffle(df)###El ajuste depende de este shuffle*****

In [3]:
print(df.shape)

(237, 513)


Se tienen 237 observaciones, cada una es una muestra en una "localización".
Por renglón, se tienen frecuencias de "bichos" de 510 "dominant taxa".

En este caso, $k=510$, $n=237$.

In [4]:
ytags=df['Tag']
print(ytags)

88         Dry forests_0
71         Dry forests_0
200         Shrublands_6
43     Cold grasslands_1
133         Grasslands_3
             ...        
123         Grasslands_2
234       Cold forests_3
79         Dry forests_0
150    Cold grasslands_5
36     Cold grasslands_1
Name: Tag, Length: 237, dtype: object


In [5]:
import numpy as np
_,idx = np.unique(ytags,return_inverse=True)

Dados los datos con su respectiva clasificación. (En el ejemplo $d=12$)

# Definición del modelo.

# Partición de datos

In [6]:
from keras.utils import to_categorical
X_total=df.drop(columns=['Tag','Dominant_taxa_ID/ID_Environmental'])
Y_total=to_categorical(idx)

In [7]:
X_train=X_total[0:200]
Y_train=Y_total[0:200]

In [8]:
df.head()
k,n=X_train.shape
n=n#-indice, -tag
print(k,n)

200 511


In [9]:
input_shape = (n,)#=n
output_shape=len(np.unique(ytags))

In [10]:
import tensorflow as tf
import keras
import pandas as pd
from keras import layers
from keras import models


In [12]:
model = models.Sequential()

model.add(layers.InputLayer(input_shape=(n,), name='Input_Layer'))#Obligatoria
model.add(layers.Dense(64, activation='relu'))# Numero de capas ocultas: Opcional
model.add(layers.Dense(32, activation='relu'))# Numero de neuronas en cada capa: Opcional
model.add(layers.Dense(37, activation='Softmax', name='Output_Layer'))#Obligatoria
model.summary()
model.compile(optimizer='adam',
              loss='categorical_crossentropy',
              metrics=['accuracy'])
import time
tic = time.time()

model.fit(x = X_train, 
           y = Y_train, 
          validation_data=[X_total[200:], Y_total[200:]],
          batch_size=50,
           epochs=30,
           verbose=2,shuffle=True)

print('seconds=', time.time()-tic)

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_2 (Dense)             (None, 64)                32768     
                                                                 
 dense_3 (Dense)             (None, 32)                2080      
                                                                 
 Output_Layer (Dense)        (None, 37)                1221      
                                                                 
Total params: 36,069
Trainable params: 36,069
Non-trainable params: 0
_________________________________________________________________
Epoch 1/30
4/4 - 2s - loss: 19.8931 - accuracy: 0.0800 - val_loss: 14.3087 - val_accuracy: 0.2432 - 2s/epoch - 409ms/step
Epoch 2/30
4/4 - 0s - loss: 10.1014 - accuracy: 0.2050 - val_loss: 9.8761 - val_accuracy: 0.2973 - 82ms/epoch - 20ms/step
Epoch 3/30
4/4 - 0s - loss: 6.1777 - accuracy: 0.3100 - val_loss: 7.22